moonraker/components/spoolman.py

```diff
@@ -0,0 +1,194 @@
+        )
+
+    async def component_init(self) -> None:
+        self.spool_id = await self.database.get_item(
+            DB_NAMESPACE, ACTIVE_SPOOL_KEY, None
+        )
+
+    async def _handle_server_ready(self):
+        self.server.register_event_handler(
+            "server:status_update", self._handle_status_update
+        )
+        result = await self.klippy_apis.subscribe_objects(
+            {"toolhead": ["position"]}
+        )
+        initial_e_pos = self._eposition_from_status(result)
+
+        logging.debug(f"Initial epos: {initial_e_pos}")
+
+        if initial_e_pos is not None:
+            self.highest_e_pos = initial_e_pos
+        else:
+            logging.error("Spoolman integration unable to subscribe to epos")
+            raise self.server.error("Unable to subscribe to e position")
+
+    def _eposition_from_status(self, status: Dict[str, Any]) -> Optional[float]:
+        position = status.get("toolhead", {}).get("position", [])
+        return position[3] if len(position) > 3 else None
+
+    async def _handle_status_update(self, status: Dict[str, Any]) -> None:
+        epos = self._eposition_from_status(status)
+        if epos and epos > self.highest_e_pos:
+            async with self.extruded_lock:
+                self.extruded += epos - self.highest_e_pos
+                self.highest_e_pos = epos
+
+            now = datetime.datetime.now()
+            difference = now - self.last_sync_time
+            if difference.total_seconds() > self.sync_rate_seconds:
+                self.last_sync_time = now
+                logging.debug("Sync period elapsed, tracking usage")
+                await self.track_filament_usage()
+
+    async def set_active_spool(self, spool_id: Optional[int]) -> None:
+        self.database.insert_item(DB_NAMESPACE, ACTIVE_SPOOL_KEY, spool_id)
+        self.spool_id = spool_id
+        await self.server.send_event(
+            "spool_manager:active_spool_set", {"spool_id": spool_id}
+        )
+        logging.info(f"Setting active spool to: {spool_id}")
+
+    def get_active_spool(self) -> Optional[int]:
+        return self.spool_id
+
+    async def track_filament_usage(self):
+        spool_id = self.get_active_spool()
+        if spool_id is None:
+            logging.debug("No active spool, skipping tracking")
+            return
+        async with self.extruded_lock:
+            if self.extruded > 0:
+                used_length = self.extruded
+
+                logging.debug(
+                    f"Sending spool usage: "
+                    f"ID: {spool_id}, "
+                    f"Length: {used_length:.3f}mm, "
+                )
+
+                response = await self.http_client.request(
+                    method="PUT",
+                    url=f"{self.spoolman_url}/spool/{spool_id}/use",
+                    body={
+                        "use_length": used_length,
+                    },
+                )
+                response.raise_for_status()
```

Comment:
I don't like fully repressing the error since then you can't debug it. I've made it now so it prints the first error since it was last successful, sounds good?